File: src/bst/BinarySearchTree.c
#include <stdio.h>
#include <stdlib.h>
#include <conio.h>

struct Node
{
    int data;
    struct Node *left, *right;
};

typedef struct Node *node;

node root = NULL;

node createNode(int val)
{
    node tmp;
    tmp = (node)malloc(sizeof(struct Node));
    tmp->data = val;
    tmp->left = NULL;
    tmp->right = NULL;
    return tmp;
}

node insertNode(node root, int x)
{
    if (root == NULL)
        return createNode(x);
    if (root->data >= x)
    {
        root->left = insertNode(root->left, x);
    }
    else if (root->data < x)
    {
        root->right = insertNode(root->right, x);
    }

    return root;
}

node search(node root, int x)
{
    if (root->data == x || root == NULL)
        return root;

    if (root->data > x)
        return search(root->left, x);

    return search(root->right, x);
}
void preOrderTraversal(node root)
{
    if (root == NULL)
        return;
    printf("%d ", root->data);
    preOrderTraversal(root->left);
    preOrderTraversal(root->right);
}

void insert(int val)
{
    root = insertNode(root, val);
}

int main()
{
    insert(5);
    insert(2);
    insert(10);
    insert(7);
    insert(1);
    insert(3);
    //preOrderTraversal(root);
    node res = search(root, 1);
    if (res == NULL)
        printf("not found");
    else
    {
        printf("%d ", res->data);
    }

    return 0;
}

File: src/bst/BinarySearchTree.java
package bst;

class BinarySearchTree {

    class Node {
        int val;
        Node left, right;

        Node(int val) {
            this.val = val;
            this.left = null;
            this.right = null;
        }
    }

    Node root;

    BinarySearchTree() {
        this.root = null;
    }

    void insert(int val) {
        root = insertValue(root, val);
    }

    Node insertValue(Node root, int val) {
        if (root == null) {
            root = new Node(val);
            return root;
        }

        if (val < root.val) {
            root.left = insertValue(root.left, val);
        } else if (val > root.val) {
            root.right = insertValue(root.right, val);
        }
        return root;
    }

    Node search(Node root, int val) {
        if (root == null || root.val == val)
            return root;

        if (val < root.val) {
            return search(root.left, val);
        }

        return search(root.right, val);
    }

    void preOrderTraversal(Node root) {
        if (root == null)
            return;

        System.out.println(root.val);
        preOrderTraversal(root.left);
        preOrderTraversal(root.right);
    }

    public static void main(String[] args) {
        BinarySearchTree bst = new BinarySearchTree();
        bst.insert(5);
        bst.insert(2);
        bst.insert(10);
        bst.insert(7);
        bst.insert(1);
        bst.insert(3);

        bst.preOrderTraversal(bst.root);

        Node res = bst.search(bst.root, 10);
        System.out.println(res.val);
    }

}